src/core/bpf-devices.c

```diff
@@ -0,0 +1,294 @@
+        if (r < 0) {
+                log_debug_errno(r, "Can't add trivial instructions to CGROUP SKB BPF program, BPF firewalling is not supported: %m");
+                return supported = BPF_DEVICES_UNSUPPORTED;
+        }
+
+        r = bpf_program_load_kernel(program, NULL, 0);
+        if (r < 0) {
+                log_debug_errno(r, "Can't load kernel CGROUP SKB BPF program, BPF firewalling is not supported: %m");
+                return supported = BPF_DEVICES_UNSUPPORTED;
+        }
+
+        attr = (union bpf_attr) {
+                .attach_type = BPF_CGROUP_DEVICE,
+                .target_fd = -1,
+                .attach_bpf_fd = -1,
+        };
+
+        if (bpf(BPF_PROG_DETACH, &attr, sizeof(attr)) < 0) {
+                if (errno != EBADF) {
+                        log_debug_errno(errno, "Didn't get EBADF from BPF_PROG_DETACH, BPF device control is not supported: %m");
+                        return supported = BPF_DEVICES_UNSUPPORTED;
+                }
+
+                /* YAY! */
+        } else {
+                log_debug("Wut? Kernel accepted our invalid BPF_PROG_DETACH call? Something is weird, assuming BPF device control is broken and hence not supported.");
+                return supported = BPF_DEVICES_UNSUPPORTED;
+        }
+
+        /* So now we know that the BPF program is generally available, let's see if BPF_F_ALLOW_MULTI is also supported
+         * (which was added in kernel 4.15). We use a similar logic as before, but this time we use the BPF_PROG_ATTACH
+         * bpf() call and the BPF_F_ALLOW_MULTI flags value. Since the flags are checked early in the system call we'll
+         * get EINVAL if it's not supported, and EBADF as before if it is available. */
+
+        attr = (union bpf_attr) {
+                .attach_type = BPF_CGROUP_DEVICE,
+                .target_fd = -1,
+                .attach_bpf_fd = -1,
+                .attach_flags = BPF_F_ALLOW_MULTI,
+        };
+
+        if (bpf(BPF_PROG_ATTACH, &attr, sizeof(attr)) < 0) {
+                if (errno == EBADF) {
+                        log_debug_errno(errno, "Got EBADF when using BPF_F_ALLOW_MULTI, which indicates it is supported. Yay!");
+                        return supported = BPF_DEVICES_SUPPORTED_WITH_MULTI;
+                }
+
+                if (errno == EINVAL)
+                        log_debug_errno(errno, "Got EINVAL error when using BPF_F_ALLOW_MULTI, which indicates it's not supported.");
+                else
+                        log_debug_errno(errno, "Got unexpected error when using BPF_F_ALLOW_MULTI, assuming it's not supported: %m");
+
+                return supported = BPF_DEVICES_SUPPORTED;
+        } else {
+                log_debug("Wut? Kernel accepted our invalid BPF_PROG_ATTACH+BPF_F_ALLOW_MULTI call? Something is weird, assuming BPF device control is broken and hence not supported.");
+                return supported = BPF_DEVICES_UNSUPPORTED;
+        }
+}
```

Comment:
the attach/detach bits can be dropped i figure, no? also the BPF_DEVICES_SUPPORTED_WITH_MULTI thing can be dropped completely, no?

for the devices bpf stuff we are in a more convenient situation than with the firewalling, as the BPF_F_ALLOW_MULTI stuff should be supported from day #1, no? Moreover, old kernels still allowed allocation of a BPF firewall program when BPF firewalling was turned off in the kernel, but then refuse attaching. I figure for bpf device control programs that bug never existed, right, and allocation of such programs was refused right-away from day #1 if bpf device control is turned off in the kernel? because of that the attchment checks here can go away entirely, no?

In fact, I figure the whole enum BPF_DEVICES_UNSUPPORTED/BPF_DEVICES_SUPPORTED/BPF_DEVICES_SUPPORTED_WITH_MULTI can just be removed altogether, as 
bpf_devices_supported() could just return 0/1 if things are supported or not...

Reply:
Yeah, you're right. Let me clean it up.